sql/migrate/migrate.go

```diff
@@ -127,303 +123,140 @@ func Schema(s *schema.Schema) StateReader {
 }
 
 type (
-	// Dir represents a versioned migration directory.
-	Dir struct {
-		fs        fs.FS
-		conn      Driver
-		pattern   string
-		templates []struct{ N, T *template.Template }
-	}
+	// FS describes the methods needed to work with the Planner.
+	FS interface {
+		// List all present migration files. Usually *.sql files.
+		List() ([]string, error)
 
-	// DirOption allows configuring the Dir
-	// using functional options.
-	DirOption func(*Dir) error
-)
+		// Read the contents of a file by name.
+		Read(name string) ([]byte, error)
 
-// NewDir creates a new workspace directory based
-// on the given configuration options.
-func NewDir(opts ...DirOption) (*Dir, error) {
-	d := &Dir{
-		fs: dirFS{
-			dir: "migrations",
-			FS:  os.DirFS("migrations"),
-		},
-		pattern: "*.sql",
-	}
-	for _, opt := range opts {
-		if err := opt(d); err != nil {
-			return nil, err
-		}
-	}
-	if len(d.templates) == 0 {
-		d.templates = []struct{ N, T *template.Template }{defaultTemplate}
+		// Write the contents to the file by name.
+		Write(name string, data []byte, perm fs.FileMode) error
+
+		// Remove a file by name.
+		Remove(name string) error
 	}
-	return d, nil
-}
 
-type (
-	// dirFS wraps the os.DirFS with additional writing capabilities.
-	dirFS struct {
-		fs.FS
-		dir string
+	// Printer wraps the methods for naming and dumping a migration plan to one or more files.
+	Printer interface {
+		// Print prints the given Plan.
+		// The returned map contains the contents for each file to dump.
+		Print(*Plan) (map[string][]byte, error)
 	}
```

Comment:
I would expect a Printer to print, Writer to write, but it looks like this one "formats" the output. Maybe use `Formater` instead?

Also, can we use `[]File` instead of a map? Iterating over a map can be annoying if you care about the order, and the `File` can be easily extended in the future.